README.md
![node](https://img.shields.io/node/v/drover-express-router.svg?style=flat-square)
![npm](https://img.shields.io/npm/v/drover-express-router.svg?style=flat-square)
![npm](https://img.shields.io/npm/dt/drover-express-router.svg?style=flat-square)
![Travis (.org)](https://img.shields.io/travis/LCMApps/drover-express-router.svg?style=flat-square)
[![Coverage Status](https://coveralls.io/repos/github/LCMApps/drover-express-router/badge.svg?branch=master)](https://coveralls.io/github/LCMApps/drover-express-router?branch=master)

# Drover express router plugin

Simple [drover](https://github.com/LCMApps/drover) plugin for [express](https://github.com/expressjs/express)
that provides convenient drover control flow via HTTP

# Installation

Using npm:

```shell
$ npm i --save drover-express-router
```

Using yarn:

```shell
$ yarn add drover-express-router
```

# Usage

```javascript
const { MasterFactory } = require('drover');

const app = require('express')();
const drover = require('drover-express-router');

const demo = MasterFactory.create({
  script: 'path/to/demo.js',
});

// one line to plug-in drover controls into your express app
app.use('/demo', drover({ master: demo }));

app.listen(3000);
```

# API

## Get status

```bash
curl -XGET localhost:3000/demo/status
//{"status": STATUS}
```

## Get workers statuses

```bash
curl -XGET localhost:3000/demo/status/workers
//{"workers": [STATUS]}
```

## Scale

```bash
curl -XPUT localhost:3000/demo/scale -d 'size=4'
```

## Graceful shutdown

```bash
curl -XPUT localhost:3000/demo/shutdown
```

## Start

```bash
curl -XPUT localhost:3000/demo/start
```

## Graceful reload

```bash
curl -XPUT localhost:3000/demo/reload
```

# Composite apps

If you got composite drover application with 2 or more masters you still can do this like this:

```javascript
const { MasterFactory } = require('drover');

const app = require('express')();
const drover = require('drover-express-router');

const foo = MasterFactory.create({ script: 'path/to/foo.js' });
const bar = MasterFactory.create({ script: 'path/to/bar.js' });

app.use('/foo', drover({ master: foo }));
app.use('/bar', drover({ master: bar }));

app.listen(3000);
```


CHANGELOG.md
## Changelog

### 1.0.0

- [#00001](https://your-issue-tracker.com/issues/00001) Initial release


tests/Unit/index.js
'use strict';

const drover = require('../../index');
const supertest = require('supertest');
const sinon = require('sinon');
const express = require('express');
const bodyParser = require('body-parser');
const { Master } = require('drover');
const { assert } = require('chai');

class AlreadyInitializedError {}
class InappropriateConditionError {}
class InvalidArgumentError {}
class WorkerStatusError {}

describe('Drover express router plugin', () => {
  it('must failed without "master" option', () => {
    const app = express();
    assert.throws(
      () => {
        app.use('/app', drover({}));
      },
      Error,
      '"master" option is required',
    );
  });

  it('must failed with invalid "master" option', () => {
    const app = express();
    assert.throws(
      () => {
        app.use('/app', drover({ master: new Object(42) }));
      },
      Error,
      '"master" option must be valid instance of "Master"',
    );
  });

  it('GET /app/status must return valid master status', async () => {
    const master = sinon.createStubInstance(Master);
    const expectedStatus = 3;
    master.getStatus.returns(expectedStatus);

    const app = express();
    app.use('/app', drover({ master }));

    const result = await supertest(app).get('/app/status');

    const expectedStatusCode = 200;

    assert.strictEqual(result.statusCode, expectedStatusCode);
    assert.strictEqual(result.body.status, expectedStatus);
  });

  it('GET /app/status must return unknown response error', async () => {
    const master = sinon.createStubInstance(Master);
    master.getStatus.throws(new Error('wtf'));

    const app = express();
    app.use('/app', drover({ master }));

    const result = await supertest(app).get('/app/status');

    const expectedStatusCode = 500;
    const expectedErrorCode = 1000;

    assert.strictEqual(result.statusCode, expectedStatusCode);
    assert.strictEqual(result.body.error.code, expectedErrorCode);
  });

  it('GET /app/status/workers must return valid master status', async () => {
    const master = sinon.createStubInstance(Master);
    const expectedStatuses = [3, 3, 3, 3];
    master.getWorkersStatuses.returns(expectedStatuses);

    const app = express();
    app.use('/app', drover({ master }));

    const result = await supertest(app).get('/app/status/workers');

    const expectedStatusCode = 200;

    assert.strictEqual(result.statusCode, expectedStatusCode);
    assert.deepEqual(result.body.workers, expectedStatuses);
  });

  it('GET /app/status/workers must return unknown response error', async () => {
    const master = sinon.createStubInstance(Master);
    master.getWorkersStatuses.throws(new Error('wtf'));

    const app = express();
    app.use('/app', drover({ master }));

    const result = await supertest(app).get('/app/status/workers');

    const expectedStatusCode = 500;
    const expectedErrorCode = 1000;

    assert.strictEqual(result.statusCode, expectedStatusCode);
    assert.strictEqual(result.body.error.code, expectedErrorCode);
  });

  it('PUT /app/start must start application and return valid master status', async () => {
    const master = sinon.createStubInstance(Master);
    const expectedStatus = 3;
    master.getStatus.returns(expectedStatus);

    const app = express();
    app.use('/app', drover({ master }));

    const result = await supertest(app).put('/app/start');

    const expectedStatusCode = 200;

    assert.ok(master.start.calledOnce);
    assert.ok(master.start.calledBefore(master.getStatus));
    assert.strictEqual(result.statusCode, expectedStatusCode);
    assert.strictEqual(result.body.status, expectedStatus);
  });

  it('PUT /app/start must return AlreadyInitializedError response', async () => {
    const master = sinon.createStubInstance(Master);
    master.start.throws(new AlreadyInitializedError());

    const app = express();
    app.use('/app', drover({ master }));

    const result = await supertest(app).put('/app/start');

    const expectedStatusCode = 400;
    const expectedErrorCode = 1001;

    assert.strictEqual(result.statusCode, expectedStatusCode);
    assert.strictEqual(result.body.error.code, expectedErrorCode);
  });

  it('PUT /app/start must return unknown response error', async () => {
    const master = sinon.createStubInstance(Master);
    master.start.throws(new Error('wtf'));

    const app = express();
    app.use('/app', drover({ master }));

    const result = await supertest(app).put('/app/start');

    const expectedStatusCode = 500;
    const expectedErrorCode = 1000;

    assert.strictEqual(result.statusCode, expectedStatusCode);
    assert.strictEqual(result.body.error.code, expectedErrorCode);
  });

  it('PUT /app/scale must scale application and return valid workers statuses', async () => {
    const master = sinon.createStubInstance(Master);
    const expectedStatuses = [3, 3, 3, 3];
    const size = 4;

    master.getWorkersStatuses.returns(expectedStatuses);

    const app = express();
    app.use(bodyParser.json());
    app.use('/app', drover({ master }));

    const result = await supertest(app)
      .put('/app/scale')
      .send({ size });

    const expectedStatusCode = 200;

    assert.ok(master.rescale.calledOnce);
    assert.ok(master.rescale.calledBefore(master.getWorkersStatuses));
    assert.deepEqual([size], master.rescale.getCall(0).args);
    assert.strictEqual(result.statusCode, expectedStatusCode);
    assert.deepEqual(result.body.workers, expectedStatuses);
  });

  it('PUT /app/scale must return unknown response error', async () => {
    const master = sinon.createStubInstance(Master);
    master.rescale.throws(new Error('wtf'));

    const app = express();
    app.use(bodyParser.json());
    app.use('/app', drover({ master }));

    const result = await supertest(app)
      .put('/app/scale')
      .send({ size: 4 });

    const expectedStatusCode = 500;
    const expectedErrorCode = 1000;

    assert.strictEqual(result.statusCode, expectedStatusCode);
    assert.strictEqual(result.body.error.code, expectedErrorCode);
  });

  it('PUT /app/scale must return InappropriateConditionError response', async () => {
    const master = sinon.createStubInstance(Master);
    master.rescale.throws(new InappropriateConditionError());

    const app = express();
    app.use(bodyParser.json());
    app.use('/app', drover({ master }));

    const result = await supertest(app)
      .put('/app/scale')
      .send({ size: 4 });

    const expectedStatusCode = 400;
    const expectedErrorCode = 1002;

    assert.strictEqual(result.statusCode, expectedStatusCode);
    assert.strictEqual(result.body.error.code, expectedErrorCode);
  });

  it('PUT /app/scale must return InvalidArgumentError response', async () => {
    const master = sinon.createStubInstance(Master);
    master.rescale.throws(new InvalidArgumentError());

    const app = express();
    app.use(bodyParser.json());
    app.use('/app', drover({ master }));

    const result = await supertest(app)
      .put('/app/scale')
      .send({ size: 4 });

    const expectedStatusCode = 400;
    const expectedErrorCode = 1003;

    assert.strictEqual(result.statusCode, expectedStatusCode);
    assert.strictEqual(result.body.error.code, expectedErrorCode);
  });

  it('PUT /app/shutdown must graceful shutdown application and return valid status', async () => {
    const master = sinon.createStubInstance(Master);
    const expectedStatus = 1;

    master.getStatus.returns(expectedStatus);

    const app = express();
    app.use(bodyParser.json());
    app.use('/app', drover({ master }));

    const result = await supertest(app).put('/app/shutdown');

    const expectedStatusCode = 200;

    assert.ok(master.gracefulShutdown.calledOnce);
    assert.ok(master.gracefulShutdown.calledBefore(master.getStatus));
    assert.strictEqual(result.statusCode, expectedStatusCode);
    assert.strictEqual(result.body.status, expectedStatus);
  });

  it('PUT /app/shutdown must return unknown response error', async () => {
    const master = sinon.createStubInstance(Master);
    master.gracefulShutdown.throws(new Error('wtf'));

    const app = express();
    app.use(bodyParser.json());
    app.use('/app', drover({ master }));

    const result = await supertest(app).put('/app/shutdown');

    const expectedStatusCode = 500;
    const expectedErrorCode = 1000;

    assert.strictEqual(result.statusCode, expectedStatusCode);
    assert.strictEqual(result.body.error.code, expectedErrorCode);
  });

  it('PUT /app/shutdown must return InappropriateConditionError response', async () => {
    const master = sinon.createStubInstance(Master);
    master.gracefulShutdown.throws(new InappropriateConditionError());

    const app = express();
    app.use(bodyParser.json());
    app.use('/app', drover({ master }));

    const result = await supertest(app).put('/app/shutdown');

    const expectedStatusCode = 400;
    const expectedErrorCode = 1002;

    assert.strictEqual(result.statusCode, expectedStatusCode);
    assert.strictEqual(result.body.error.code, expectedErrorCode);
  });

  it('PUT /app/shutdown must return WorkerStatusError response', async () => {
    const master = sinon.createStubInstance(Master);
    master.gracefulShutdown.throws(new WorkerStatusError());

    const app = express();
    app.use(bodyParser.json());
    app.use('/app', drover({ master }));

    const result = await supertest(app).put('/app/shutdown');

    const expectedStatusCode = 500;
    const expectedErrorCode = 1004;

    assert.strictEqual(result.statusCode, expectedStatusCode);
    assert.strictEqual(result.body.error.code, expectedErrorCode);
  });

  it('PUT /app/reload must graceful reload application and return valid worker statuses', async () => {
    const master = sinon.createStubInstance(Master);
    const expectedStatuses = [3, 3, 3, 3];

    master.getWorkersStatuses.returns(expectedStatuses);

    const app = express();
    app.use(bodyParser.json());
    app.use('/app', drover({ master }));

    const result = await supertest(app).put('/app/reload');

    const expectedStatusCode = 200;

    assert.ok(master.gracefulReload.calledOnce);
    assert.ok(master.gracefulReload.calledBefore(master.getWorkersStatuses));
    assert.strictEqual(result.statusCode, expectedStatusCode);
    assert.deepEqual(result.body.workers, expectedStatuses);
  });

  it('PUT /app/reload must return unknown response error', async () => {
    const master = sinon.createStubInstance(Master);
    master.gracefulReload.throws(new Error('wtf'));

    const app = express();
    app.use(bodyParser.json());
    app.use('/app', drover({ master }));

    const result = await supertest(app).put('/app/reload');

    const expectedStatusCode = 500;
    const expectedErrorCode = 1000;

    assert.strictEqual(result.statusCode, expectedStatusCode);
    assert.strictEqual(result.body.error.code, expectedErrorCode);
  });

  it('PUT /app/reload must return InappropriateConditionError response', async () => {
    const master = sinon.createStubInstance(Master);
    master.gracefulReload.throws(new InappropriateConditionError());

    const app = express();
    app.use(bodyParser.json());
    app.use('/app', drover({ master }));

    const result = await supertest(app).put('/app/reload');

    const expectedStatusCode = 400;
    const expectedErrorCode = 1002;

    assert.strictEqual(result.statusCode, expectedStatusCode);
    assert.strictEqual(result.body.error.code, expectedErrorCode);
  });

  it('PUT /app/reload must return WorkerStatusError response', async () => {
    const master = sinon.createStubInstance(Master);
    master.gracefulReload.throws(new WorkerStatusError());

    const app = express();
    app.use(bodyParser.json());
    app.use('/app', drover({ master }));

    const result = await supertest(app).put('/app/reload');

    const expectedStatusCode = 500;
    const expectedErrorCode = 1004;

    assert.strictEqual(result.statusCode, expectedStatusCode);
    assert.strictEqual(result.body.error.code, expectedErrorCode);
  });
});


index.js
'use strict';

const { Router } = require('express');

const UnknownErrorResponse = { error: { code: 1000, message: 'Unknown error' } };
const AlreadyStartedErrorResponse = { error: { code: 1001, message: 'Service already started' } };
const InappropriateConditionErrorResponse = { error: { code: 1002, message: 'Inappropriate conditions' } };
const InvalidScaleSizeErrorResponse = { error: { code: 1003, message: 'Invalid scale size' } };
const WorkerStatusErrorResponse = {
  error: { code: 1004, message: 'One or more workers have error state' },
};

/**
 * @param {Master} options.master
 * @return {createApplication.Router}
 */
module.exports = options => {
  const { master } = options;

  if (!master) {
    throw new Error('"master" option is required');
  }

  if (master.constructor.name !== 'Master') {
    throw new Error('"master" option must be valid instance of "Master"');
  }

  const router = new Router();

  router.get('/status', (req, res) => {
    try {
      res.send({ status: master.getStatus() });
    } catch (err) {
      res.status(500).send(UnknownErrorResponse);
    }
  });

  router.get('/status/workers', (req, res) => {
    try {
      res.send({ workers: master.getWorkersStatuses() });
    } catch (err) {
      res.status(500).send(UnknownErrorResponse);
    }
  });

  router.put('/start', async (req, res) => {
    try {
      await master.start();

      res.send({ status: master.getStatus() });
    } catch (err) {
      if (err.constructor.name === 'AlreadyInitializedError') {
        return res.status(400).send(AlreadyStartedErrorResponse);
      }

      res.status(500).send(UnknownErrorResponse);
    }
  });

  router.put('/scale', async (req, res) => {
    const { size } = req.body;

    try {
      await master.rescale(size);

      res.send({ workers: master.getWorkersStatuses() });
    } catch (err) {
      if (err.constructor.name === 'InappropriateConditionError') {
        return res.status(400).send(InappropriateConditionErrorResponse);
      } else if (err.constructor.name === 'InvalidArgumentError') {
        return res.status(400).send(InvalidScaleSizeErrorResponse);
      }

      res.status(500).send(UnknownErrorResponse);
    }
  });

  router.put('/shutdown', async (req, res) => {
    try {
      await master.gracefulShutdown();

      res.send({ status: master.getStatus() });
    } catch (err) {
      if (err.constructor.name === 'InappropriateConditionError') {
        return res.status(400).send(InappropriateConditionErrorResponse);
      } else if (err.constructor.name === 'WorkerStatusError') {
        return res.status(500).send(WorkerStatusErrorResponse);
      }

      res.status(500).send(UnknownErrorResponse);
    }
  });

  router.put('/reload', async (req, res) => {
    try {
      await master.gracefulReload();

      res.send({ workers: master.getWorkersStatuses() });
    } catch (err) {
      if (err.constructor.name === 'InappropriateConditionError') {
        return res.status(400).send(InappropriateConditionErrorResponse);
      } else if (err.constructor.name === 'WorkerStatusError') {
        return res.status(500).send(WorkerStatusErrorResponse);
      }

      res.status(500).send(UnknownErrorResponse);
    }
  });

  return router;
};
